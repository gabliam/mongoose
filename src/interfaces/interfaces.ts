import * as mongoose from 'mongoose';

export interface DocumentOptions  {
  name: string;
  collectionName?: string;
  schema?: mongoose.Schema;
}

export interface IRead<T extends mongoose.Document> {
  findAll(): Promise<T[]>;
  findById(_id: string): Promise<T>;
  findOne(cond?: Object): mongoose.Query<T>;
  find(cond?: Object, fields?: Object, options?: Object): mongoose.Query<T[]>;
}

export interface IWrite<T, U extends mongoose.Document> {
  create(item: T): Promise<U>;
  update(_id: mongoose.Types.ObjectId, item: T): Promise<U>;
  delete(_id: string): Promise<void>;
}