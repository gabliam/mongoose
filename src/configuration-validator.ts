import * as Joi from 'joi';

export const configurationValidator = Joi.object()
  .keys({
    uri: Joi.string(),
    host: Joi.string(),
    database_name: Joi.string(),
    port: Joi.number().default(27017),
    options: Joi.object()
  })
  .xor('uri', 'host')
  .and('host', 'database_name');
