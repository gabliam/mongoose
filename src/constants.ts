/** Type in registry */
export const TYPE = {
  Document: 'Document',
};

export const METADATA_KEY = {
  document: '_document'
};

export const LIST_DOCUMENT = Symbol('LIST_DOCUMENT');
