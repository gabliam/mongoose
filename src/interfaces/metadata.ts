import * as mongoose from 'mongoose';

export interface DocumentMetadata {
    name: string;
    collectionName?: string;
    schema: mongoose.Schema;
}