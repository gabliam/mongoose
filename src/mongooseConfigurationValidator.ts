import * as Joi from 'joi';

export const mongooseConfigurationValidator = Joi.object().keys({
  uri: Joi.string(),
  host: Joi.string(),
  database_name: Joi.string(),
  port: Joi.number(),
  options: Joi.object(),
}).xor('uri', 'host').and('host', 'database_name');
