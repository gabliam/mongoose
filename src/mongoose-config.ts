import { PluginConfig, Value, Bean, inject } from '@gabliam/core';
import { LIST_DOCUMENT } from './constants';
import { configurationValidator } from './configuration-validator';
import { MongooseConfiguration } from './interfaces';
import { MongooseConnection } from './mongoose-connection';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:mongoose');

@PluginConfig()
export class PluginMongooseConfig {

  @Value('application.mongoose', configurationValidator)
  mongooseConfiguration: MongooseConfiguration;

  listDocument: any[];

  constructor(
    @inject(LIST_DOCUMENT) listDocument: any[]
  ) {
    debug('constructor PluginMongooseConfig');
    this.listDocument = listDocument;
  }

  @Bean(MongooseConnection)
  create() {
    debug('Create MongooseConnection', this.mongooseConfiguration);
    return new MongooseConnection(this.mongooseConfiguration, this.listDocument);
  }
}
