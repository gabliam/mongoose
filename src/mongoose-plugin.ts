import {
  interfaces as coreInterfaces,
  inversifyInterfaces,
  Scan,
  Registry,
  Plugin
} from '@gabliam/core';
import { TYPE, LIST_DOCUMENT } from './constants';
import * as d from 'debug';
const debug = d('Gabliam:Plugin:mongoose');
import { MongooseConnection } from './mongoose-connection';

@Plugin()
@Scan()
export class MongoosePlugin implements coreInterfaces.GabliamPlugin {
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    const documents = registry.get(TYPE.Document).map(({ id, target }) => {
      return target;
    });
    debug('list documents', documents);
    container.bind<any>(LIST_DOCUMENT).toConstantValue(documents);
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    const connection = container.get<MongooseConnection>(MongooseConnection);
    if (connection) {
      await connection.conn.close();
    }
  }
}
