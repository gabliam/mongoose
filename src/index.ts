import * as interfaces from './interfaces';
import { MongoosePlugin } from './mongoose-plugin';

export * from './decorators';
export { interfaces };
export * from './mongoose-connection';
export * from './repository';
export { MongoosePlugin as default };
