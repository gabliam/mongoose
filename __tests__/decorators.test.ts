import { RegistryMetada } from '@gabliam/core/lib/interfaces';
import { Document } from '../src';
import { METADATA_KEY as CORE_METADATA_KEY } from '@gabliam/core/lib/constants';
import { DocumentMetadata } from '../src/interfaces/index';
import { METADATA_KEY } from '../src/constants';
import * as mongoose from 'mongoose';

test(`@Document('Hero') decorator`, () => {
  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true
        }
      });
    }
  }

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    Hero
  );

  const documentMetadata: DocumentMetadata = Reflect.getOwnMetadata(
    METADATA_KEY.document,
    Hero
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
});

test(`@Document(DocumentOptions without collectionName)  decorator`, () => {
  @Document({
    name: 'Hero',
    schema: new mongoose.Schema({
      name: {
        type: String,
        required: true
      }
    })
  })
  class Hero {}

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    Hero
  );

  const documentMetadata: DocumentMetadata = Reflect.getOwnMetadata(
    METADATA_KEY.document,
    Hero
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
});

test(`@Document(DocumentOptions)  decorator`, () => {
  @Document({
    name: 'Hero',
    collectionName: 'HeroCollection',
    schema: new mongoose.Schema({
      name: {
        type: String,
        required: true
      }
    })
  })
  class Hero {}

  const entityMetadata: RegistryMetada = Reflect.getMetadata(
    CORE_METADATA_KEY.register,
    Hero
  );

  const documentMetadata: DocumentMetadata = Reflect.getOwnMetadata(
    METADATA_KEY.document,
    Hero
  );

  expect(entityMetadata).toMatchSnapshot();
  expect(documentMetadata).toMatchSnapshot();
});

describe('errors', () => {
  test('no getSchema', () => {
    expect(() => {
      @Document('Hero')
      class Hero {}
      // tslint:disable-next-line:no-unused-expression
      new Hero();
    }).toThrowError();
  });

  test('getSchema return bad info', () => {
    expect(() => {
      @Document('Hero')
      class Hero {
        static getSchema() {
          return 'lol';
        }
      }
      // tslint:disable-next-line:no-unused-expression
      new Hero();
    }).toThrowError();
  });
});
