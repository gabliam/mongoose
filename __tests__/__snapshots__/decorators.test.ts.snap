// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Document('Hero') decorator 1`] = `
Object {
  "type": "Document",
  "value": Object {
    "id": null,
    "target": [Function],
  },
}
`;

exports[`@Document('Hero') decorator 2`] = `
Object {
  "name": "Hero",
  "schema": Schema {
    "_indexes": Array [],
    "aliases": Object {},
    "callQueue": Array [
      Array [
        "pre",
        Arguments [
          "remove",
          true,
          [Function],
        ],
      ],
    ],
    "childSchemas": Array [],
    "inherits": Object {},
    "methods": Object {},
    "nested": Object {},
    "obj": Object {
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "options": Object {
      "_id": true,
      "autoIndex": null,
      "bufferCommands": true,
      "capped": false,
      "discriminatorKey": "__t",
      "id": true,
      "minimize": true,
      "noId": false,
      "noVirtualId": false,
      "read": null,
      "retainKeyOrder": false,
      "shardKey": null,
      "strict": true,
      "typeKey": "type",
      "validateBeforeSave": true,
      "versionKey": "__v",
    },
    "paths": Object {
      "_id": ObjectId {
        "_index": null,
        "defaultValue": [Function],
        "getters": Array [],
        "instance": "ObjectID",
        "options": Object {
          "auto": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "path": "_id",
        "setters": Array [
          [Function],
        ],
        "validators": Array [],
      },
      "name": SchemaString {
        "_index": null,
        "enumValues": Array [],
        "getters": Array [],
        "instance": "String",
        "isRequired": true,
        "options": Object {
          "required": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "originalRequiredValue": true,
        "path": "name",
        "regExp": null,
        "requiredValidator": [Function],
        "setters": Array [],
        "validators": Array [
          Object {
            "message": "Path \`{PATH}\` is required.",
            "type": "required",
            "validator": [Function],
          },
        ],
      },
    },
    "plugins": Array [],
    "query": Object {},
    "s": Object {
      "hooks": Kareem {
        "_posts": Object {},
        "_pres": Object {},
      },
      "kareemHooks": Object {
        "count": true,
        "find": true,
        "findOne": true,
        "findOneAndRemove": true,
        "findOneAndUpdate": true,
        "insertMany": true,
        "update": true,
        "updateMany": true,
        "updateOne": true,
      },
    },
    "singleNestedPaths": Object {},
    "statics": Object {},
    "subpaths": Object {},
    "tree": Object {
      "_id": Object {
        "auto": true,
        "type": [Function],
      },
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "virtuals": Object {},
  },
}
`;

exports[`@Document(DocumentOptions without collectionName)  decorator 1`] = `
Object {
  "type": "Document",
  "value": Object {
    "id": null,
    "target": [Function],
  },
}
`;

exports[`@Document(DocumentOptions without collectionName)  decorator 2`] = `
Object {
  "name": "Hero",
  "schema": Schema {
    "_indexes": Array [],
    "aliases": Object {},
    "callQueue": Array [
      Array [
        "pre",
        Arguments [
          "remove",
          true,
          [Function],
        ],
      ],
    ],
    "childSchemas": Array [],
    "inherits": Object {},
    "methods": Object {},
    "nested": Object {},
    "obj": Object {
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "options": Object {
      "_id": true,
      "autoIndex": null,
      "bufferCommands": true,
      "capped": false,
      "discriminatorKey": "__t",
      "id": true,
      "minimize": true,
      "noId": false,
      "noVirtualId": false,
      "read": null,
      "retainKeyOrder": false,
      "shardKey": null,
      "strict": true,
      "typeKey": "type",
      "validateBeforeSave": true,
      "versionKey": "__v",
    },
    "paths": Object {
      "_id": ObjectId {
        "_index": null,
        "defaultValue": [Function],
        "getters": Array [],
        "instance": "ObjectID",
        "options": Object {
          "auto": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "path": "_id",
        "setters": Array [
          [Function],
        ],
        "validators": Array [],
      },
      "name": SchemaString {
        "_index": null,
        "enumValues": Array [],
        "getters": Array [],
        "instance": "String",
        "isRequired": true,
        "options": Object {
          "required": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "originalRequiredValue": true,
        "path": "name",
        "regExp": null,
        "requiredValidator": [Function],
        "setters": Array [],
        "validators": Array [
          Object {
            "message": "Path \`{PATH}\` is required.",
            "type": "required",
            "validator": [Function],
          },
        ],
      },
    },
    "plugins": Array [],
    "query": Object {},
    "s": Object {
      "hooks": Kareem {
        "_posts": Object {},
        "_pres": Object {},
      },
      "kareemHooks": Object {
        "count": true,
        "find": true,
        "findOne": true,
        "findOneAndRemove": true,
        "findOneAndUpdate": true,
        "insertMany": true,
        "update": true,
        "updateMany": true,
        "updateOne": true,
      },
    },
    "singleNestedPaths": Object {},
    "statics": Object {},
    "subpaths": Object {},
    "tree": Object {
      "_id": Object {
        "auto": true,
        "type": [Function],
      },
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "virtuals": Object {},
  },
}
`;

exports[`@Document(DocumentOptions)  decorator 1`] = `
Object {
  "type": "Document",
  "value": Object {
    "id": null,
    "target": [Function],
  },
}
`;

exports[`@Document(DocumentOptions)  decorator 2`] = `
Object {
  "collectionName": "HeroCollection",
  "name": "Hero",
  "schema": Schema {
    "_indexes": Array [],
    "aliases": Object {},
    "callQueue": Array [
      Array [
        "pre",
        Arguments [
          "remove",
          true,
          [Function],
        ],
      ],
    ],
    "childSchemas": Array [],
    "inherits": Object {},
    "methods": Object {},
    "nested": Object {},
    "obj": Object {
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "options": Object {
      "_id": true,
      "autoIndex": null,
      "bufferCommands": true,
      "capped": false,
      "discriminatorKey": "__t",
      "id": true,
      "minimize": true,
      "noId": false,
      "noVirtualId": false,
      "read": null,
      "retainKeyOrder": false,
      "shardKey": null,
      "strict": true,
      "typeKey": "type",
      "validateBeforeSave": true,
      "versionKey": "__v",
    },
    "paths": Object {
      "_id": ObjectId {
        "_index": null,
        "defaultValue": [Function],
        "getters": Array [],
        "instance": "ObjectID",
        "options": Object {
          "auto": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "path": "_id",
        "setters": Array [
          [Function],
        ],
        "validators": Array [],
      },
      "name": SchemaString {
        "_index": null,
        "enumValues": Array [],
        "getters": Array [],
        "instance": "String",
        "isRequired": true,
        "options": Object {
          "required": true,
          "runSettersOnQuery": undefined,
          "type": [Function],
        },
        "originalRequiredValue": true,
        "path": "name",
        "regExp": null,
        "requiredValidator": [Function],
        "setters": Array [],
        "validators": Array [
          Object {
            "message": "Path \`{PATH}\` is required.",
            "type": "required",
            "validator": [Function],
          },
        ],
      },
    },
    "plugins": Array [],
    "query": Object {},
    "s": Object {
      "hooks": Kareem {
        "_posts": Object {},
        "_pres": Object {},
      },
      "kareemHooks": Object {
        "count": true,
        "find": true,
        "findOne": true,
        "findOneAndRemove": true,
        "findOneAndUpdate": true,
        "insertMany": true,
        "update": true,
        "updateMany": true,
        "updateOne": true,
      },
    },
    "singleNestedPaths": Object {},
    "statics": Object {},
    "subpaths": Object {},
    "tree": Object {
      "_id": Object {
        "auto": true,
        "type": [Function],
      },
      "name": Object {
        "required": true,
        "type": [Function],
      },
    },
    "virtuals": Object {},
  },
}
`;
